src/Orleans/Core/GrainFactory.cs

```diff
@@ -37,10 +25,28 @@ public class GrainFactory : IGrainFactory
         private readonly ConcurrentDictionary<Type, IGrainMethodInvoker> invokers =
             new ConcurrentDictionary<Type, IGrainMethodInvoker>();
 
+        /// <summary>
+        /// The cache of typed system target references.
+        /// </summary>
+        private readonly Dictionary<Tuple<GrainId, Type>, Dictionary<SiloAddress, ISystemTarget>> typedSystemTargetReferenceCache =
+                    new Dictionary<Tuple<GrainId, Type>, Dictionary<SiloAddress, ISystemTarget>>();
+
+        /// <summary>
+        /// The cache of type metadata.
+        /// </summary>
+        private readonly TypeMetadataCache typeCache;
+
+        /// <summary>
+        /// The runtime client.
+        /// </summary>
+        private readonly IRuntimeClient runtimeClient;
+
         // Make this internal so that client code is forced to access the IGrainFactory using the 
         // GrainClient (to make sure they don't forget to initialize the client).
-        internal GrainFactory()
+        public GrainFactory(IRuntimeClient runtimeClient, TypeMetadataCache typeCache)
```

Comment:
Having GrainFactory depending on IRuntimeClient seems backwards, as like I mentioned IRuntimeClient IMO should just be a facade on top of other services. I would depend on the other services directly from here. Also, I would not let InsideRuntimeClient create and expose the IGrainFactory service, it should just get it from the provider, but create it somewhere else. Ideally (at some point in the future) there shouldn't be a need for our code to depend on IRuntimeClient, and only external code might use it as a simplified facade (realistically I don't know how practical that is given the current shape of the runtime client)